Python/bytecodes.c

```diff
@@ -3945,16 +3957,22 @@ dummy_func(
 
         op(_SET_IP, (--)) {
             TIER_TWO_ONLY
-            frame->prev_instr = ip_offset + oparg;
+            frame->instr_ptr = ip_offset + oparg;
         }
 
         op(_SAVE_CURRENT_IP, (--)) {
-            TIER_ONE_ONLY
-            frame->prev_instr = next_instr - 1;
+            #if TIER_ONE
+            assert(frame->next_instr_offset == 0);
+            frame->next_instr_offset = (uint16_t)(next_instr - frame->instr_ptr);
+            #endif
+            #if TIER_TWO
```

Comment:
Why is this being added this back to tier 2? It should be eliminated by the tier 2 optimizer.

Reply:
Without this addition some tests failed (possibly only when run with `-Xuops`). I don't recall the details but unfortunately this was the best I could do. As I [wrote](https://github.com/python/cpython/pull/109095#discussion_r1362929287) when I was debugging this for Irit, it seems we've lost track of what exactly the VM state should be and how to maintain it in Tier 2. I think you're right when you suggested in the same comment thread that we should rename this uop to `_SAVE_RETURN_OFFSET`. Also note that there are exactly two occurrences of `_SAVE_CURRENT_IP`: in the two `CALL_...` macros, immediately before `_PUSH_FRAME`.

Reply:
As I said above, I think the problem is that `next_instr` is part of the state when transferring control from tier2 to tier 1.
We cannot, in general, determine `next_instr` from `frame->instr_ptr`.
We can, however, determine it in both the tier1 and tier 2 interpreters because we know how long the instruction is.
`next_instr = frame->instr_ptr + len(instruction)` It is during the transfer from tier2 to tier1 that we don't know `next_instr` as we don't know `len(instruction)`
Ergo, we need to return `next_instr` from executors.